libs/sqlserver/langchain_sqlserver/vectorstores.py

```diff
@@ -225,31 +229,86 @@ def __init__(
+            if "=" in arg:
+                # Split into key value pairs by the first positioned `=` found.
+                # Key-Value pairs are inserted into the dictionary.
+                #
+                key, value = arg.split("=", 1)
+                arg_dict[key.lower().strip()] = value.strip()
 
-        Check the components of the connection string to determine
-        if connection via Entra ID authentication is possible or not.
+        try:
+            # This will throw a key error if server or database keyword
+            # is not present in arg_dict from the connection string.
+            #
+            database = arg_dict.pop("database")
 
-        The connection string is of expected to be of the form:
-            "mssql+pyodbc://username:password@servername/dbname?other_params"
-        which gets parsed into -> <scheme>://<netloc>/<path>?<query>
-        """
-        parsed_url = urlparse(self.connection_string)
+            # If `server` is present in the dictionary, we split by
+            # `,` to obtain host and port details.
+            #
+            server = arg_dict.pop("server").split(",", 1)
+            server_host = server[0]
+            server_port = None
 
-        if parsed_url is None:
-            logging.error("Unable to parse connection string.")
-            return False
+            # Check if port is provided in server details,if true,
+            # cast value to int if possible.
+            #
+            if len(server) > 1 and server[1].isdigit():
+                server_port = int(server[1])
 
-        if (parsed_url.username and parsed_url.password) or (
-            "trusted_connection=yes" in parsed_url.query.lower()
+        except KeyError as k:
+            logging.error(
+                f"Server, DB details were not provided in the connection string.\n{k}"
+            )
+            raise Exception(
+                "Server, DB details should be provided in connection string."
+            )
+
+        # Args needed to be checked
+        #
+        username = arg_dict.pop("uid", None)
+        password = arg_dict.pop("pwd", None)
+
+        if "driver" in arg_dict.keys():
+            # Extract driver value from curly braces if present.
+            driver = re.search(r"\{([^}]*)\}", arg_dict["driver"])
+            if driver is not None:
+                arg_dict["driver"] = driver.group(1)
+
+        # Determine if Entra ID connection is possible
+        #
+        if (username and password) or (
+            "trustedconnection" in arg_dict.keys()
+            and arg_dict["trustedconnection"].lower() == "yes"
         ):
-            return False
+            self._can_connect_with_entra_id = False
+
+        # Create connection URL for SQLAlchemy
+        #
+        url = URL.create(
+            "mssql+pyodbc",
+            username=username,
+            password=password,
+            database=database,
+            host=server_host,
+            port=server_port,
```

Comment:
if server_port is not specified, will this send in a None? if yes, is that fine for the sqlalchemy connection string creation?

Reply:
Yes. Port value is optional. We can connect without specifying it.